### PR TITLE
Fix/mismatched toggle optimizer

### LINKED ISSUE #7405: Step optimizers at arbitrary intervals
## 🐛 Bug

In my model I have two optimizers that alternate during the optimize steps as shown below:

```python
class PLModel(LightningModule):

    def __init__(self, hparams):

        super(PLModel, self).__init__()
        self.hparams = hparams
        
        ...


    def configure_optimizers(self):

        # optimizers
        optimizers = [
            torch.optim.AdamW(self.x1_encoder.parameters(), lr=self.hparams.lr, betas=(0.9, 0.999), eps=1e-08,
                             weight_decay=self.hparams.weight_decay, amsgrad=True),
            torch.optim.AdamW(self.x2_encoder.parameters(), lr=self.hparams.lr, betas=(0.9, 0.999), eps=1e-08,
                             weight_decay=self.hparams.weight_decay, amsgrad=True)
        ]

        # schedulers
        step_size_up = 0.03 * self.num_training_steps

        schedulers = [
            torch.optim.lr_scheduler.CyclicLR(
                optimizers[0],
                mode='triangular2',
                base_lr=self.hparams.base_lr,
                max_lr=self.hparams.max_lr,
                step_size_up=step_size_up,
                cycle_momentum=False),
            torch.optim.lr_scheduler.CyclicLR(
                optimizers[1],
                mode='triangular2',
                base_lr=self.hparams.base_lr,
                max_lr=self.hparams.max_lr,
                step_size_up=step_size_up,
                cycle_momentum=False)
        ]
        return optimizers, schedulers

    # Alternating schedule for optimizer steps
    def optimizer_step(
            self, epoch, batch_idx, optimizer, optimizer_idx, optimizer_closure,
            on_tpu, using_native_amp, using_lbfgs
    ):
        # update x1_encoder opt
        if optimizer_idx == 0:
            if batch_idx % 2 == 0:
                optimizer.step(closure=optimizer_closure)

        # update x2_encoder opt
        if optimizer_idx == 1:
            if batch_idx % 2 != 0:
                optimizer.step(closure=optimizer_closure)
```

But I am getting the following error message during the optimizer step:

```python
Epoch 0:   0% 0/4086 [00:07<?, ?it/s]
Traceback (most recent call last):
  File "main.py", line 213, in perform_tasks
    fit(hparams)
  File "main.py", line 89, in fit
    trainer.fit(model, datamodule=dm)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/trainer.py", line 637, in run_train
    self.train_loop.run_training_epoch()
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 492, in run_training_epoch
    batch_output = self.run_training_batch(batch, batch_idx, dataloader_idx)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 654, in run_training_batch
    self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 433, in optimizer_step
    using_lbfgs=is_lbfgs,
  File "/content/source/model/PLModel.py", line 82, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/core/optimizer.py", line 214, in step
    self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/core/optimizer.py", line 134, in __optimizer_step
    trainer.accelerator.optimizer_step(optimizer, self._optimizer_idx, lambda_closure=closure, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/accelerators/accelerator.py", line 274, in optimizer_step
    self.lightning_module, optimizer, opt_idx, lambda_closure, **kwargs
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/precision/native_amp.py", line 78, in pre_optimizer_step
    lambda_closure()
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 649, in train_step_and_backward_closure
    split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 755, in training_step_and_backward
    self.backward(result, optimizer, opt_idx)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/trainer/training_loop.py", line 785, in backward
    result.closure_loss, optimizer, opt_idx, should_accumulate, *args, **kwargs
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/accelerators/accelerator.py", line 257, in backward
    self.lightning_module, closure_loss, optimizer, optimizer_idx, should_accumulate, *args, **kwargs
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/precision/native_amp.py", line 59, in backward
    closure_loss = super().backward(model, closure_loss, optimizer, opt_idx, should_accumulate, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 71, in backward
    model.backward(closure_loss, optimizer, opt_idx)
  File "/usr/local/lib/python3.7/dist-packages/pytorch_lightning/core/lightning.py", line 1251, in backward
    loss.backward(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/tensor.py", line 245, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/usr/local/lib/python3.7/dist-packages/torch/autograd/__init__.py", line 147, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

Could anyone tell me what I'm missing?

### To Reproduce

This issue could be reproduced using BoringModel on [Google Colab](https://colab.research.google.com/drive/1HYUCYo3TQDO3b57ZvHkGmreqhzdkTaAD?usp=sharing).

### Expected behavior
Without override `optimizer_step` method, training happens normally. Therefore, applying step optimizers at arbitrary intervals theoretically should occur normally.

**OBS**: when using `pytorch-lightning==1.1.3`, `optimizer_step` was running smoothly

### Environment

``` python
* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu101
	- pytorch-lightning: 1.2.10
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
```

### PR DESCRIPTION
**Pull request by Lucklyric**
## What does this PR do?

Fixes #7405

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- NA Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — CHANGES_REQUESTED** [review]
Can you resolve conflicts and add a test?

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- Changed calling of `untoggle_optimizer(opt_idx)` out of the closure function
```
Comment (SkafteNicki) [inline CHANGELOG.md:24]:
```suggestion
- Changed calling of `untoggle_optimizer(opt_idx)` out of the closure function ([#7563](https://github.com/PyTorchLightning/pytorch-lightning/pull/7563)
```

Comment (Lucklyric) [reply CHANGELOG.md]: Updated

**Comment by Lucklyric** [comment]
Dear @carmocca , I just resolved the conflict. Because I only changed the place of `untoggle_optimizer`,  those codes are still covered by existing tests. The Codcov reports 0 changes on `training_loop.py`. I am actually not that sure why it shows 5% decreasing from many other files.

**Comment by carmocca** [comment]
> I just resolved the conflict

Thanks!

> those codes seem to be still covered by existing tests.

Coverage is not enough. Needs a test to ensure the new behaviour is correct and to avoid future regressions

**Comment by tchaton** [comment]
Dear @Lucklyric,

Mind making a test triggering the grad_fn error on master and working there.

Best,
T.C

**Comment by Lucklyric** [comment]
> Dear @Lucklyric,
> 
> Mind making a test triggering the grad_fn error on master and working there.
> 
> Best,
> T.C

Will do that. Thank you!

**Comment by Lucklyric** [comment]
Dear @carmocca @tchaton , added a test that fails in master and works here. I'm not sure if this is a proper test, so please feel free to provide feedback. Thanks

**Comment by tchaton** [comment]
Hey @Lucklyric @carmoccam confirm this test is failing on master.
Great job there !

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 2]
```diff
@@ -168,3 +168,70 @@ def training_step(self, batch, batch_idx):
 
     with pytest.raises(ValueError, match='`training_step` is missing the `optimizer_idx`'):
         trainer.fit(TestModel())
+
+
+def test_csutom_optimizer_step_with_multiple_optimizers(tmpdir):
```
Comment (ananthsub) [inline tests/trainer/optimization/test_multiple_optimizers.py:173]:
```suggestion
def test_custom_optimizer_step_with_multiple_optimizers(tmpdir):
```

**Review by ananthsub — APPROVED** [review]

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 3]
```diff
@@ -168,3 +168,70 @@ def training_step(self, batch, batch_idx):
...
+    trainer = pl.Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=4,
+        max_epochs=1,
+        log_every_n_steps=1,
+        weights_summary=None,
+    )
+    trainer.fit(model)
+    assert all([a == b for a, b in zip(model.training_step_called, [4, 2])])
+    assert all([a == b for a, b in zip(model.optimizer_step_called, [4, 2])])
```
Comment (carmocca) [inline tests/trainer/optimization/test_multiple_optimizers.py:237]:
```suggestion
    assert model.training_step_called == [4, 2]
    assert model.optimizer_step_called == [4, 2]
```

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 4]
```diff
@@ -168,3 +168,70 @@ def training_step(self, batch, batch_idx):
...
+        def optimizer_step(
+            self,
+            epoch,
+            batch_idx,
+            optimizer,
+            optimizer_idx,
+            optimizer_closure,
+            on_tpu=False,
+            using_native_amp=False,
+            using_lbfgs=False,
```
Comment (carmocca) [inline tests/trainer/optimization/test_multiple_optimizers.py:212]:
```suggestion
            *_,
```

**Review by carmocca — APPROVED** [review]
Some nits. Thanks!

**Comment by Lucklyric** [comment]
> Some nits. Thanks!

Thanks for your comments, @carmocca . Updated!

**Pull request merged**